### DISCUSSION TITLE
bash正则表达式匹配

### DISCUSSION THREAD

**Discussion by DM0U** [physics-data/faq#99] [Q&A] [answered]
在bash中，以下语句不能完成匹配
`[[ "Content-Length: 14" =~ ^Content-Length:' '\d+ ]]`
或
`[[ "Content-Length: 14" =~ ^Content-Length:' '\w+ ]]`

但是
`[[ "Content-Length: 14" =~ ^Content-Length:' '14 ]]`
能成功匹配，同时其它语句中\w也能成功匹配。
为什么这句话中\w和\d不能实现匹配？

**Comment by YiHui-Liu** [comment, marked as answer]
![image](https://user-images.githubusercontent.com/30190271/131671673-77a88ccb-dd9c-450f-8f14-3f60f143b08a.png)
使用`'Content-Length: \w+'`匹配上了

**Reply by YiHui-Liu** [reply]
![image](https://user-images.githubusercontent.com/30190271/131671952-0fb4bc5e-9bb1-4453-87c0-7c3eaaaaee40.png)
尝试之后，结果如上

<hr />

正则表达式有`BREs`、`EREs`、`PREs`等不同的标准，默认情况下bash里支持的是`BREs`，而这一个标准是不支持`\d`匹配的
另外，空格不用打单引号，当然也不能用`\s`，这个同样不支持，直接使用` `就可以了

**Reply by heroxbd** [reply]
太强了， @DM0U 楼上正解。

**Comment by heroxbd** [comment]
> 在bash中，以下语句不能完成匹配
> `[[ "Content-Length: 14" =~ ^Content-Length:' '\d+ ]]`
> 或
> `[[ "Content-Length: 14" =~ ^Content-Length:' '\w+ ]]`
> 
> 但是
> `[[ "Content-Length: 14" =~ ^Content-Length:' '14 ]]`
> 能成功匹配，同时其它语句中\w也能成功匹配。
> 为什么这句话中\w和\d不能实现匹配？

可以用 `[0-9]` 替代 `\d`
